3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,285 @@
+      .map(PiiScalar(_).success)
+      .getOrElse(s"The specified pojo field $fieldName is not supported".failure)
+
+  private def extractPiiJsonField(jsonField: JValue): Validation[String, PiiJson] = {
+    val schemaCriterion = extractString(jsonField, "schemaCriterion")
+      .flatMap(sc => SchemaCriterion.parse(sc).leftMap(_.getMessage))
+      .toValidationNel
+    val jsonPath = extractString(jsonField, "jsonPath").toValidationNel
+    val mutator = extractString(jsonField, "field")
+      .flatMap(getJsonMutator)
+      .toValidationNel
+    val validatedNel = (mutator |@| schemaCriterion |@| jsonPath)(PiiJson.apply)
+    validatedNel.leftMap(x => s"Unable to extract PII JSON: ${x.list.mkString(",")}")
+  }
+
+  private def getJsonMutator(fieldName: String): Validation[String, Mutator] =
+    JsonMutators
+      .get(fieldName)
+      .map(_.success)
+      .getOrElse(s"The specified json field ${fieldName} is not supported".failure)
+
+  private def extractString(jValue: JValue, field: String, tail: String*): Validation[String, String] =
+    ScalazJson4sUtils.extract[String](jValue, field, tail: _*).leftMap(_.getMessage)
+
+  private def extractStrategyFunction(config: JValue): Validation[String, String] =
+    ScalazJson4sUtils
+      .extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction")
+      .leftMap(_.getMessage)
+
+  private def matchesSchema(config: JValue, schemaKey: SchemaKey): Validation[String, JValue] =
+    if (supportedSchema.matches(schemaKey)) {
+      config.success
+    } else {
+      "Schema key %s is not supported. A '%s' enrichment must have schema '%s'."
+        .format(schemaKey, supportedSchema.name, supportedSchema)
+        .failure
+    }
+}
+
+/**
+ * Implements a pseudonymization strategy using any algorithm known to MessageDigest
+ * @param hashFunction the MessageDigest function to apply
+ */
+final case class PiiStrategyPseudonymize(hashFunction: MessageDigest) extends PiiStrategy {
+  val TextEncoding                                 = "UTF-8"
+  override def scramble(clearText: String): String = hash(clearText)
+  def hash(text: String): String = {
+    val bytes = hashFunction.digest(text.getBytes(TextEncoding))
+    hashFunction.reset()
+    String.format("%0" + bytes.size * 2 + "x", new java.math.BigInteger(1, bytes))
+  }
+}
+
+/**
+ * The PiiPseudonymizerEnrichment runs after all other enrichments to find fields that are configured as PII (personally
+ * identifiable information) and apply some anonymization (currently only pseudonymization) on them. Currently a single
+ * strategy for all the fields is supported due to the configuration format, and there is only one implemented strategy,
+ * however the enrichment supports a strategy per field.
+ *
+ * The user may specify two types of fields POJO or JSON. A POJO field is effectively a scalar field in the
+ * EnrichedEvent, whereas a JSON is a "context" formatted field and it can be wither a scalar in the case of
```

Comment:
can be a scalar

Reply:
you mean for the types. Actually in the config they are called pojo and json. I can add the clarification.

Reply:
no I meant the "wither" typo

Reply:
Oh I see, it was supposed to be `either`